### PR TITLE
Allow to interact on some outside elements #42

### PR DESCRIPTION
**Pull request by gejgalis**
Please let me know if it is acceptable. Maybe configuration name should be different?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -61,6 +61,7 @@ Returns a new focus trap on `element`.
 - **escapeDeactivates** {boolean}: Default: `true`. If `false`, the `Escape` key will not trigger deactivation of the focus trap. This can be useful if you want to force the user to make a decision instead of allowing an easy way out.
 - **clickOutsideDeactivates** {boolean}: Default: `false`. If `true`, a click outside the focus trap will deactivate the focus trap and allow the click event to do its thing.
```
Comment (davidtheclark) [inline README.md]: @gejgalis: Thanks for your patience. I think I understand your goals here and agree that it's useful. I have one source of hesitation: I worry that the relationship of `clickOutsideDeactivates` and `includeOutsideElement` is kind of confusing. What do you think about changing this up a little so we do not add an `includeOutsideElement` option but instead allow `clickOutsideDeactivates` to be a function? These would be the rules for `clickOutsideDeactivates`:

- If `true`, all clicks outside the focus trap deactivate it.
- If `false`, no clicks outside the focus trap deactivate it.
- If a function, when there's a click outside the focus trap the clicked element is passed to the function and a boolean is returned. If the function returns `true`, the focus trap is deactivated. If the function returns `false, the focus trap is not.

I'm thinking the advantage is that we no longer have a conflict between two options: `clickOutsideDeactivates` is the only one you set to control what outside clicks do.

How does that sound?

Comment (gejgalis) [reply README.md]: I'm not sure if your proposal solves my problem. I don't want to activate/deactivate focus trap when clicked outside - I just want to skip preventing click events outside.

I think your previous proposal was better: istead of adding new option `includeOutsideElement` we could allow to setup multiple elements? What do you think @davidtheclark ?

Comment (davidtheclark) [reply README.md]: > I'm not sure if your proposal solves my problem.

@gejgalis I don't understand how it doesn't? — maybe the weirdness of the Kendo UI is confusing me. I thought the goal is to allow you to click on (but not focus) *some* external elements (the datepicker) without deactivating the focus trap? If so, I think it makes sense to say, essentially, "clickOutsideDeactivates when you click anywhere on the screen except within the datepicker" — and that's what a filter function could allow for.

Comment (gejgalis) [reply README.md]: @davidtheclark I'm convinced now. It makes sense and it should work for me. You want me to prepare PR?

Comment (davidtheclark) [reply README.md]: Glad it made sense to you. Yeah, a PR would be great, @gejgalis!

**Comment by pgn-vole** [comment]
Can this PR be looked at again? We need the same feature, for React Components not kendo ones though. 
I think the initial implementaion makes more sense that your new proposed one @davidtheclark. One doesn't want to deactivate the focus trap entirely but simply let some node to not be ignored when the focus trap is active.

**Comment by davidtheclark** [comment]
@pgn-vole You're welcome to open another PR following through on this plan. I don't think I understand your objection, though.

> the goal is to allow you to click on (but not focus) some external elements (the datepicker) without deactivating the focus trap

Is that not your goal?

**Comment by pgn-vole** [comment]
@davidtheclark That's exactly my goal. 
Your suggestion, to re-use the API `clickOutsideDeactivates` does more than just this though. It entirely disable the focus trap when clicking on an outside element. We don't need that, we just would like for some clicks to pass through while keeping the focus trap enable.
Does that make sense?

**Comment by davidtheclark** [comment]
> Your suggestion, to re-use the API clickOutsideDeactivates does more than just this though. It entirely disable the focus trap when clicking on an outside element.

No: my suggestion is to make `clickOutsideDeactivates` accept a function that you can use to determine whether a click on an outside element disables the focus trap or not.

**Comment by pgn-vole** [comment]
I see, so they are actually two different use cases :)
I will work on another PR to solve the problem I'm facing then.

**Comment by davidtheclark** [comment]
Closing: see https://github.com/davidtheclark/focus-trap/issues/42#issuecomment-408714255

**Pull request closed without merging**